docs/cvl/cvl2/changes.md

```diff
@@ -0,0 +1,583 @@
+}
+```
+
+In CVL 2, you must now use the name of the contract, rather than the variable,
+when referring to user-defined types.  The above example would now be written
+
+```cvl
+rule example {
+    Example.S x = getAnS();
+}
+```
+
+There is no need for a `using` statement in this example.
+
+`using` statements are still required to call methods on secondary contracts.
+
+```{todo}
+Error message
+```
+
+Changes to methods block entries
+--------------------------------
+
+In addition to the superficial changes listed above, there are some changes that
+change the way that methods block entries can be written.  In CVL 1, `methods`
+block entries often had several different functions and meanings:
+
+ - They are used to indicate targets for summarization
+ - They are used to write generic specs that could apply to contracts with
+   missing methods
+ - They are used to declare targets `envfree`
+
+With these changes, these different uses are more explicit.
+
+(cvl2-visibility)=
+### Required `internal` or `external` annotation
+
+Every methods block entry must be marked either `internal` or `external`.  The
+annotation must come after the argument list and before the `returns` clause.
+
+If a function is declared `public` in Solidity, then the Solidity compiler
+creates an internal implementation method, and an external wrapper method that
+calls the internal implementation.  Therefore, you can summarize a `public`
+method by marking the summarization `internal`.
+
+If you want to summarize both the internal implementation and the external
+wrapper, you need to add two separate entries to the `methods` block.
+
+```{todo}
+If you do not change this, you will see the following error:
+```
+
+### `optional` methods entries
+
+In CVL 1, you could write an entry in the methods block for a method that does
+not exist in the contract; rules that would call the non-existent method are
+skipped during verification.
+
+This behavior can lead to confusion, because typos or name changes could silently
+cause a rule to be skipped.
+
+In CVL 2, this behavior is still available, but the methods entry must contain
+the keyword `optional` somewhere after the `returns` clause and before the
+summarization (if any).
+
+```{todo}
+If a methods block contains a non-optional entry for a method that doesn't exist
+in the contract, you will receive the following error message:
+```
+
+### `library` annotations
+
+In CVL 2, contract functions declared as library functions must be annotated
+with `library` in the `methods` block.
+
+```{todo}
+If you forget to declare a method as a `library` method, you will receive the
+following error message:
+```
+
+### Required `calldata`, `memory`, or `storage` annotations for reference types
+
+In CVL 2, methods entries for internal functions must contain either `calldata`,
+`memory`, or `storage` annotations for all arguments with reference types (such
+as arrays).
+
+```{todo}
+is `calldata` actually one of the options?
```

Comment:
It is definitely possible to write an internal function with a parameter in a `calldata` location